docs/design_notes/module_lifecycle.md

```diff
@@ -0,0 +1,289 @@
+
+### Compact control flow
+
+The order in which you define submodules determines the name of a submodule if none is provided explicitly (using the name= construction argument). Because the name determines how parameters are mapped to submodules, you must be careful about mixing control flow with auto-generated names. Using control flow can change the order or remove certain submodules altogether. This is useful in the case where certain submodules should only exist depending on some construction argument. However, when control flow depends on the input arguments to the Module you should be careful. For example the following Module will break:
+
+
+```python
+class WrongModule(nn.Module):
+  @nn.compact
+  def __call__(self, x, mode):
+    if mode == "encode"
+      return nn.Dense(8)(x)
+    elif mode == "decode":
+      return nn.Dense(4)(x)
+```
+
+
+This problem can be solved in various ways: Provide explicit names, create the modules in setup, or move the constructor out of the control flow. The latter is done as follows:
+
+
+```python
+class CorrectModule(nn.Module):
+  @nn.compact
+  def __call__(self, x, mode):
+    encoder = nn.Dense(8)
+    decoder = nn.Dense(4)
+    if mode == "encode"
+      return encoder(x)
+    elif mode == "decode":
+      return decoder(x)
+```
+
+
+In the above example the construction order is fixed. After construction the submodules can be used in an arbitrary order.
+
+> Note \
+compact modules show a strong resemblance to [React hooks](https://reactjs.org/docs/hooks-custom.html)
+
+
+## Top-level Modules
+
+When a Module instance is created at the top-level it will be in an "unbound" state, meaning it has no variables attached. By top-level we mean it is not constructed as a sub-Module inside another Module class. Apart from calling `init` and `apply` there is not much you can do with an unbound Module. Note also that `setup` is not called on unbound Modules so you can really only access the construction arguments. See the Future work section for how this might change in the future.
+
+### Why are top-level modules always unbound?
+
+When we call `apply`, a copy of the top-level Module is created which will actually hold the variables and PRNG sequences. This stateful, "bound", clone only exists while we are executing the apply method. The reason for this is that if you create a stateful object and destroy it before the apply function returns, the `apply` function itself behaves like a pure function (pure means if you put the same arguments in, it will return the same outputs).
+
+Pure functions have many advantages but when using JAX they are often essential. For example, most code requires compilation using `jax.jit` to be fast and once you created a Module you probably want to optimize its parameters using `jax.grad`. However, these APIs expect a pure function and don't work on stateful bound `Module` instances directly. Moreover, pure functions allow for flexible interoperability with other libraries.
```

Comment:
> Moreover, pure functions allow for flexible interoperability with other libraries.

Can you give an example? Maybe mention Optax?

Reply:
done